File: 2023-04(백준, 프로그래머스)/수식 최대화.py
from itertools import permutations
from collections import deque
from copy import deepcopy

def solution(ex):
    op = []
    for i in ['-','+','*']:
        if i in ex:
            op.append(i)
            ex = ex.replace(i,f',{i},')
    ex = ex.split(',')
    
    result = 0
    for items in permutations(op,len(op)):
        q = deque(deepcopy(ex))
        for item in items:
            check = []
            while q:
                n = q.popleft()
                if n == item and n == '-':
                    check.append(str(int(check.pop()) - int(q.popleft())))
                elif n == item and n == '+':
                    check.append(str(int(check.pop()) + int(q.popleft())))
                elif n == item and n == '*':
                    check.append(str(int(check.pop()) * int(q.popleft())))
                else:
                    check.append(n)
            q = deque(check)
        result = max(result, abs(int(q.pop())))
    return result

# 파이썬에는 eval이라는 함수가 있다.
# eval('5+5') = 10 // 수식이 문자열로 이루어져도 실행가능하게 해준다.(알고있으면 좋을듯 하다.)

File: JS 코테 연습/수식 최대화.js
function solution(ex) {
    let op = []
    
    for (let i =0; i < ex.length; i++){
        if (ex[i] === '-' || ex[i] === '+' || ex[i] =='*'){
            op.push(ex[i])
        }
    }
    op = [...new Set(op)]
    
    for (i of op){
        ex = ex.replaceAll(i, `,${i},`)
    }
    ex = ex.split(',')

    if (op.length === 3){
        op = [['-', '+', '*'],['-', '*', '+'],['*', '+', '-'],['*', '-', '+'],['+', '-', '*'],['+', '*', '-']]
    }else if (op.length === 2){
        op = [op, [op[1], op[0]]]
    }
    
    // JS에서 무한대 값 넣기
    let result = -Infinity
    for (i of op){
        // map매서드를 이용한 깊은 복사
        let copy = [...Array(ex.length).keys()].map(n => ex[n])
        for (j of i){
            let check = []
            while (copy.length !== 0){
                let p = copy.shift()
                if (p === j && p === '+'){
                    check.push(Number(check.pop()) + Number(copy.shift()))
                }else if(p === j && p === '-'){
                    check.push(Number(check.pop()) - Number(copy.shift()))
                }else if(p === j && p === '*'){
                    check.push(Number(check.pop()) * Number(copy.shift()))
                }else{
                    check.push(p)
                }
            }
            copy = check
        }
        result = Math.max(result, Math.abs(copy[0]))
    }
    
    return result
}
solution("100-200*300-500+20")